### PR TITLE
Introduce @http:CacheConfig annotation to the resource

### PR DESCRIPTION
**Pull request by TharmiganK**
## Purpose

> Fixes [`ballerina-standard-library/issues/1533`](https://github.com/ballerina-platform/ballerina-standard-library/issues/1533)

This annotation can be used to enable response caching from the resource signature. This allows to set the `cache-control`, `etag` and `last-modified` headers in the response. 

The default behavior (`@http:CacheConfig`) is to have `must-revalidate,public,max-age=3600` directives in `cache-control` header. In addition to that `etag` and `last-modified` headers will be added. 

```ballerina
@http:CacheConfig {           // Default Configuration
    mustRevalidate : true,    // Sets the must-revalidate directive
    noCache : false,          // Sets the no-cache directive
    noStore : false,          // Sets the no-store directive 
    noTransform : false,      // Sets the no-transform directive
    isPrivate : false,        // Sets the private and public directive
    proxyRevalidate : false,  // Sets the proxy-revalidate directive
    maxAge : 600,             // Sets the max-age directive. Default value is 3600 seconds
    sMaxAge : -1,             // Sets the s-maxage directive
    noCacheFields : [],       // Optional fields for no-cache directive
    privateFields : [],       // Optional fields for private directive
    setETag : true,           // Sets the etag header
    setLastModified : true    // Sets the last-modified header
}
```
This annotation can **only** support return types of `anydata` and `SucessStatusCodeResponse`. (For other return values cache configuration will not be added through this annotation)

## Example

```ballerina
// Sets the cache-control header as "public,must-revalidate,max-age=5". Also sets the etag header.
// last-modified header will not be set
resource function get cachingBackEnd(http:Request req) returns @http:CacheConfig{maxAge : 5, 
    setLastModified : false} string {

    return "Hello, World!!"
}
```

## Checklist
- [x] Linked to an issue
- [x] Updated the changelog
- [x] Added tests

### CODE REVIEW

File: ballerina-tests/tests/http_cache_config_annotation_test.bal [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+            res.statusCode = 500;
+            res.setPayload(response.message());
+            checkpanic caller->respond(res);
+        }
+    }
+}
+
+service / on new http:Listener(cacheAnnotationTestPort2) {
+
+    resource function default nocacheBE(http:Request req) returns @http:CacheConfig{} json {
```
[the remark this answers is not in the document]
Comment (TharmiganK) [reply ballerina-tests/tests/http_cache_config_annotation_test.bal:99]: Yes

File: ballerina/http_annotation.bal [hunk 2]
```diff
@@ -96,3 +96,39 @@ public type HttpHeader record {|
...
+    decimal maxAge = -1;
+    decimal sMaxAge = -1;
+    string[] noCacheFields = [];
+    string[] privateFields = [];
+    boolean setETag = true;
+    boolean setLastModified = true;
+|};
+
+# The annotation which is used to define the response cache configuration. This annotation only supports `anydata` and
+# `SuccessStatusCodeResponse` return types.
```
Comment (chamil321) [inline ballerina/http_annotation.bal:133]:
```suggestion
# Success(2XX) `StatusCodeResponses` return types.
```

File: ballerina/http_connection.bal [hunk 3]
```diff
@@ -113,9 +113,11 @@ public client class Caller {
         return nativeGetRemoteHostName(self);
     }
 
-    private isolated function returnResponse(anydata|StatusCodeResponse|Response|error message, string? returnMediaType)
-            returns ListenerError? {
+    private isolated function returnResponse(anydata|StatusCodeResponse|Response|error message, string? returnMediaType,
+        HttpCacheConfig? cacheConfig) returns ListenerError? {
         Response response = new;
+        boolean setETag = cacheConfig is () ? false: cacheConfig.setETag;
+        boolean enableCache = false;
```
Comment (chamil321) [inline ballerina/http_connection.bal:120]:
```suggestion
        boolean cacheCompatibleType = false;
```
Is this more readable? So we know that, this flag is there to show the compatible types

File: ballerina-tests/tests/http_cache_config_annotation_test.bal [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+
+    resource function get maxAge(http:Caller caller, http:Request req) {
+        http:Response|error response = cacheBackendEP->forward("/maxAgeBE", req);
+        if (response is http:Response) {
+            checkpanic caller->respond(response);
+        } else {
+            http:Response res = new;
+            res.statusCode = 500;
+            res.setPayload(response.message());
+            checkpanic caller->respond(res);
```
Comment (chamil321) [inline ballerina-tests/tests/http_cache_config_annotation_test.bal:65]: Try to use StatusCodeResponse here also.

File: ballerina-tests/tests/http_cache_config_annotation_test.bal [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+string mustRevalidatePayload1 = "Hello, World!";
+byte[] mustRevalidatePayload2 = "Hello, New World!".toBytes();
+json nocachePayload1 = { "message": "1st response" };
+json nocachePayload2 = { "message": "2nd response" };
+http:Ok ok = {body : mustRevalidatePayload1};
+http:InternalServerError err = {body : errorBody};
+
+service / on new http:Listener(cacheAnnotationTestPort1) {
+
+    resource function get noCache(http:Caller caller, http:Request req) {
```
Comment (chamil321) [inline ballerina-tests/tests/http_cache_config_annotation_test.bal:44]: Avoid caller and return http:Response

File: ballerina-tests/tests/http_cache_config_annotation_test.bal [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+json nocachePayload1 = { "message": "1st response" };
+json nocachePayload2 = { "message": "2nd response" };
+http:Ok ok = {body : mustRevalidatePayload1};
+http:InternalServerError err = {body : errorBody};
+
+service / on new http:Listener(cacheAnnotationTestPort1) {
+
+    resource function get noCache(http:Request req) returns http:Response|http:InternalServerError {
+        http:Response|error response = cacheBackendEP->forward("/nocacheBE", req);
+        if (response is http:Response) {
```
Comment (lahirusamith) [inline ballerina-tests/tests/http_cache_config_annotation_test.bal:45]: parentheses are not required around the condition so we can remove parentheses around if condition. 

 https://docs.google.com/document/d/1y6QVqaZzZt9jMpYV4jP5WRS_W_KoC4y40Uuoh1ALu8E/edit#

File: ballerina-tests/tests/http_cache_config_annotation_test.bal [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+    if (response is http:Response) {
+        test:assertEquals(response.statusCode, 200, msg = "Found unexpected output");
+        test:assertEquals(noCacheHitCountNew, 1);
+        test:assertTrue(response.hasHeader(LAST_MODIFIED));
+        assertHeaderValue(checkpanic response.getHeader(CACHE_CONTROL), "no-cache,public");
+        assertHeaderValue(checkpanic response.getHeader(ETAG), crypto:crc32b(nocachePayload1.toString().toBytes()));
+        assertHeaderValue(checkpanic response.getHeader(CONTENT_TYPE), APPLICATION_JSON);
+        assertJsonPayload(response.getJsonPayload(), nocachePayload1);
+    } else {
+        test:assertFail(msg = "Found unexpected output type: " + response.message());
```
Comment (lahirusamith) [inline ballerina-tests/tests/http_cache_config_annotation_test.bal:140]: we can use `check` to leave out error but need to change the function returns as below. 

```ballerina
function testNoCacheCacheControlWithAnnotation() returns error? {
    http:Response response = check cacheClientEP->get("/noCache");
}
```

File: ballerina/http_annotation.bal [hunk 8]
```diff
@@ -96,3 +96,40 @@ public type HttpHeader record {|
...
+# + setETag - Sets the `etag` header for the given payload
+# + setLastModified - Sets the current time as the `last-modified` header
+public type HttpCacheConfig record {|
+    boolean mustRevalidate = true;
+    boolean noCache = false;
+    boolean noStore = false;
+    boolean noTransform = false;
+    boolean isPrivate = false;
+    boolean proxyRevalidate = false;
+    decimal maxAge = 600;
```
Comment (shafreenAnfar) [inline ballerina/http_annotation.bal:124]: Lets increase it to 1 hour.

**Review by shafreenAnfar — APPROVED** [review]

**Review by chamil321 — APPROVED** [review]

**Pull request merged**